activerecord/lib/active_record/connection_adapters/mysql_adapter.rb

```diff
@@ -416,8 +392,11 @@ def exec_stmt(sql, name, binds)
             # place when an error occurs. To support older MySQL versions, we
             # need to close the statement and delete the statement from the
             # cache.
-            stmt.close
-            @statements.delete sql
+            if binds.empty?
+              stmt.close
+            else
+              @statements.delete sql
+            end
```

Comment:
Err, that comment says "and". Either this code change is wrong, or the comment needs to be fixed to match.